relations/src/shielder/withdraw.rs

```diff
@@ -155,28 +262,41 @@ impl ConstraintSynthesizer<CircuitField> for WithdrawRelation {
         //----------------------------------
         // Check the token values soundness.
         //----------------------------------
-        let token_amount_out =
-            FpVar::new_input(ns!(cs, "token amount out"), || Ok(&self.token_amount_out))?;
+        let token_amount_out = FpVar::new_input(ns!(cs, "token amount out"), || {
+            self.token_amount_out.ok_or(AssignmentMissing)
+        })?;
         // some range checks for overflows?
         let token_sum = token_amount_out.add(new_token_amount);
         token_sum.enforce_equal(&whole_token_amount)?;
 
         //------------------------
         // Check the merkle proof.
         //------------------------
-        let merkle_root = FpVar::new_input(ns!(cs, "merkle root"), || Ok(&self.merkle_root))?;
-        let mut leaf_index = FpVar::new_witness(ns!(cs, "leaf index"), || Ok(&self.leaf_index))?;
+        let merkle_root = FpVar::new_input(ns!(cs, "merkle root"), || {
+            self.merkle_root.ok_or(AssignmentMissing)
+        })?;
+        let mut leaf_index = FpVar::new_witness(ns!(cs, "leaf index"), || {
+            self.leaf_index.ok_or(AssignmentMissing)
+        })?;
 
         let mut current_hash_bytes = old_note.to_bytes()?;
-        for hash in self.merkle_path {
-            let sibling = FpVar::new_witness(ns!(cs, "merkle path node"), || Ok(hash))?;
+        let mut hash_bytes = vec![current_hash_bytes.clone()];
+        let path = self.merkle_path.unwrap_or_default();
+
+        let zero = CircuitField::zero();
+
+        for i in 0..self.max_path_len {
```

Comment:
I guess you could validate the length of `self.merkle_path <= self.max_path_len` first. If it's false then this can't be valid proof anyway. The logic here wouldn't have to change then either.

Reply:
sure, added